pkg/security/probe/probe_ebpf.go

```
@@ -1678,43 +1546,182 @@ func (p *EBPFProbe) handleEvent(CPU int, data []byte) {
-			seclog.Errorf("failed to decode prctl event: %s (offset %d, len %d)", err, offset, len(data))
-			return
+		if err = p.regularUnmarshalEvent(&event.PrCtl, eventType, offset, dataLen, data); err != nil {
+			return false
 		}
 		if event.PrCtl.IsNameTruncated {
 			p.MetricNameTruncated.Add(1)
 		}
 	}
+	return true
+}
 
-	// send related events
-	for _, relatedEvent := range relatedEvents {
-		p.DispatchEvent(relatedEvent, true)
-		p.eventPool.Put(relatedEvent)
+func (p *EBPFProbe) handleBeforeProcessContext(eventType model.EventType, event *model.Event, data []byte, offset int, dataLen uint64, newEntryCb func(entry *model.ProcessCacheEntry, err error)) bool {
+	var err error
+	switch eventType {
+	case model.ForkEventType:
+		if _, err = p.unmarshalProcessCacheEntry(event, data[offset:]); err != nil {
+			seclog.Errorf("failed to decode fork event: %s (offset %d, len %d)", err, offset, dataLen)
+			return false
+		}
+
+		if err := p.Resolvers.ProcessResolver.AddForkEntry(event, newEntryCb); err != nil {
+			seclog.Errorf("failed to insert fork event: %s (pid %d, offset %d, len %d)", err, event.PIDContext.Pid, offset, len(data))
+			return false
+		}
+	case model.ExecEventType:
+		// unmarshal and fill event.processCacheEntry
+		if _, err = p.unmarshalProcessCacheEntry(event, data[offset:]); err != nil {
+			seclog.Errorf("failed to decode exec event: %s (offset %d, len %d)", err, offset, len(data))
+			return false
+		}
+
+		err = p.Resolvers.ProcessResolver.AddExecEntry(event)
+		if err != nil {
+			seclog.Errorf("failed to insert exec event: %s (pid %d, offset %d, len %d)", err, event.PIDContext.Pid, offset, len(data))
+			return false
+		}
 	}
-	relatedEvents = relatedEvents[0:0]
+	return true
+}
 
-	p.DispatchEvent(event, true)
+func (p *EBPFProbe) handleEarlyReturnEvents(eventType model.EventType, event *model.Event, offset int, dataLen uint64, data []byte, newEntryCb func(entry *model.ProcessCacheEntry, err error)) bool {
```

could you return the error instead of a boolean ?

---

Sure, but the err is already printed inside the function so I don't know if we want to print it again